### ISSUE TITLE
Dependencies graph skips

### ISSUE THREAD

**Issue by makkes** [weaveworks/weave-gitops#3302]
**Describe the bug**

When opening an application's graph, the visualization is re-drawn differently every couple of seconds.

**Environment**

- Weave-Gitops Version: 0.15.0

**To Reproduce**

1. Go to an application that has some dependencies
1. Open the "Graph" tab
1. Wait for a couple of seconds

**Expected behavior**

The graph doesn't change.

**Actual Behavior**

The graph is re-drawn differently, making it hard to follow a certain path.

**Additional Context (screenshots, logs, etc)**

https://user-images.githubusercontent.com/387444/214308989-77e12e87-772d-45c9-8eff-ad1e99580af8.mp4

**Comment by makkes** [comment]
To add another data point I reproduced this issue with older version of Weave GitOps and saw it happening back until 0.10.1 *but not* in 0.10.0. Something that has been changed in 0.10.1 causes this.

**Comment by jpellizzari** [comment]
Root cause is that sometimes the server returns more objects on one request than another. I have added sorting logic on a branch, which works nicely for simple hierarchies, but on complex graphs (flux-system Kustomization, for example), the server returns 80 objects on one request, and 96 on another, etc.

Sorting cannot account for more objects being added to the graph.

**Comment by jpellizzari** [comment]
Some thoughts on how to fix:

1. Turn off polling (:cry:). Turn off polling on complex graphs? (also :cry: ).
2. Append-only after initial render? Might need some mutative data structures or something a bit fancier than normal to pull this off. Or maybe just concat-ing an array is enough.

**Comment by tun0** [comment]
What is "the server" in this case?

And if it's an "external" problem, how come it worked "properly" in earlier releases?

**Comment by makkes** [comment]
> Root cause is that sometimes the server returns more objects on one request than another. I have added sorting logic on a branch, which works nicely for simple hierarchies, but on complex graphs (flux-system Kustomization, for example), the server returns 80 objects on one request, and 96 on another, etc.
> 
> Sorting cannot account for more objects being added to the graph.

Shouldn't we rather fix the issue that the server seems to return inconsistent results?

**Comment by makkes** [comment]
> What is "the server" in this case?

The server is part of Weave GitOps. It's responsible for querying objects from the underlying Kubernetes cluster, authenticating requests etc.

> And if it's an "external" problem, how come it worked "properly" in earlier releases?

Apparently changes to the server-side component led to the change in behaviour as I laid out in my first comment.

**Comment by jpellizzari** [comment]
> Shouldn't we rather fix the issue that the server seems to return inconsistent results?

If that is fixable, yes.

**Comment by jpellizzari** [comment]
More info: removing the `go` keyword from this func (making it synchronous) results in consistent objects getting returned:

https://github.com/weaveworks/weave-gitops/blob/7d1d4389824634ebdaa7a95c96f448fe497a7360/core/server/fluxruntime.go#L247

So probably a race condition there. Still working on it.
